aiosmtpd/controller.py

```diff
@@ -191,3 +184,89 @@ def stop(self):
         self._thread.join()
         self._thread = None
         self._thread_exception = None
+
+
+@public
+class Controller(BaseThreadedController):
+    def __init__(
+        self,
+        handler,
+        hostname: Optional[str] = None,
+        port: int = 8025,
+        loop=None,
+        *,
+        ready_timeout: float = 1.0,
+        ssl_context: ssl.SSLContext = None,
+        # SMTP parameters
+        server_hostname: Optional[str] = None,
+        **SMTP_parameters,
+    ):
+        super().__init__(
+            handler,
+            loop,
+            ready_timeout=ready_timeout,
+            server_hostname=server_hostname,
+            **SMTP_parameters
+        )
+        self.hostname = "::1" if hostname is None else hostname
+        self.port = port
+        self.ssl_context = ssl_context
+
+    def _create_server(self) -> Coroutine:
+        return self.loop.create_server(
+            self._factory_invoker,
+            host=self.hostname,
+            port=self.port,
+            ssl=self.ssl_context,
+        )
+
+    def _trigger_server(self):
+        """
+        Opens a socket connection to the newly launched server, wrapping in an SSL
+        Context if necessary, and read some data from it to ensure that factory()
+        gets invoked.
+        """
+        with ExitStack() as stk:
+            s = stk.enter_context(create_connection((self.hostname, self.port), 1.0))
+            if self.ssl_context:
+                s = stk.enter_context(self.ssl_context.wrap_socket(s))
+            _ = s.recv(1024)
+
+
+class UnixSocketController(BaseThreadedController):  # pragma: on-win32
+    def __init__(
+        self,
+        handler,
+        unix_socket: Optional[Union[str, Path]],
+        loop=None,
+        *,
+        ready_timeout=1.0,
+        ssl_context=None,
+        # SMTP parameters
+        server_hostname: str = None,
+        **SMTP_parameters,
+    ):
+        super().__init__(
+            handler,
+            loop,
+            ready_timeout=ready_timeout,
+            ssl_context=ssl_context,
+            server_hostname=server_hostname,
+            **SMTP_parameters
+        )
+        self.unix_socket = str(unix_socket)
+
+    def _create_server(self) -> Coroutine:
+        return self.loop.create_unix_server(
+            self._factory_invoker,
+            path=self.unix_socket,
+            ssl=self.ssl_context,
+        )
+
+    def _trigger_server(self):
+        with ExitStack() as stk:
+            s: socket = stk.enter_context(socket(AF_UNIX, SOCK_STREAM))
+            s.connect(self.unix_socket)
+            if self.ssl_context:
+                s = stk.enter_context(self.ssl_context.wrap_socket(s))
+            _ = s.recv(1024)
```

Comment:
New class! The gist of Issue #114 

It's amazing how simple the diffcode is after the shared code is extracted to `BaseThreadedController`